axiom/datasets.go

```diff
@@ -379,70 +385,109 @@ func (s *DatasetsService) IngestEvents(ctx context.Context, id string, events []
 // Restrictions for field names (JSON object keys) can be reviewed here:
 // https://www.axiom.co/docs/usage/field-restrictions.
 //
-// As this method keeps a connection open until the channel is closed, it is not
-// advised to use this method for long-running ingestions.
+// Events are ingested in batches. A batch is either 1024 events for unbuffered
+// channels or the capacity of the channel for buffered channels. A batch is
+// sent to the server as soon as it is full or after one second or when the
+// channel is closed.
 func (s *DatasetsService) IngestChannel(ctx context.Context, id string, events <-chan Event, options ...ingest.Option) (*ingest.Status, error) {
+	const (
+		defaultBatchSize    = 1024
+		defaultSendInterval = time.Second
+	)
+
 	ctx, span := s.client.trace(ctx, "Datasets.IngestChannel", trace.WithAttributes(
 		attribute.String("axiom.dataset_id", id),
 		attribute.Int("axiom.channel.capacity", cap(events)),
 	))
 	defer span.End()
 
-	// Apply supplied options.
-	var opts ingest.Options
-	for _, option := range options {
-		option(&opts)
-	}
+	var (
+		errCh        = make(chan error)
+		ingestStatus ingest.Status
+	)
+	go func() {
+		defer close(errCh)
 
-	path, err := AddOptions(s.basePath+"/"+id+"/ingest", opts)
-	if err != nil {
-		return nil, spanError(span, err)
-	}
+		// Batch is either 1024 events for unbuffered channels or the capacity
+		// of the channel for buffered channels.
+		batchSize := defaultBatchSize
+		if cap(events) != 0 {
+			batchSize = cap(events)
+		}
+		batch := make([]Event, 0, batchSize)
 
-	pr, pw := io.Pipe()
+		ingestBatch := func(ctx context.Context) error {
+			if len(batch) == 0 {
+				return nil
+			}
 
-	zsw, err := zstd.NewWriter(pw)
-	if err != nil {
-		_ = pr.Close()
-		_ = pw.Close()
-		return nil, err
-	}
+			// Reset batch buffer after attempted ingest, regardless of the
+			// outcome.
+			defer func() {
+				batch = make([]Event, 0, batchSize)
+			}()
 
-	go func() {
-		var (
-			enc    = json.NewEncoder(zsw)
-			encErr error
-		)
-		for event := range events {
-			if encErr = enc.Encode(event); encErr != nil {
-				break
+			res, err := s.IngestEvents(ctx, id, batch, options...)
+			if err != nil {
+				return err
 			}
-		}
 
-		if closeErr := zsw.Close(); encErr == nil {
-			// If we have no error from encoding but from closing, capture that
-			// one.
-			encErr = closeErr
+			ingestStatus.Add(res)
+
+			return nil
 		}
-		_ = pw.CloseWithError(encErr)
-	}()
 
-	req, err := s.client.NewRequest(ctx, http.MethodPost, path, pr)
-	if err != nil {
-		return nil, spanError(span, err)
-	}
+		t := time.NewTicker(defaultSendInterval)
+		defer t.Stop()
 
-	req.Header.Set("Content-Type", NDJSON.String())
-	req.Header.Set("Content-Encoding", Zstd.String())
+		for {
+			select {
+			case <-ctx.Done():
+				flushCtx, flushCancel := context.WithTimeout(context.Background(), 10*time.Second)
+				defer flushCancel()
```

Comment:
looks dubious - as it could stack up an unlimited amount of contexts... deferred functions are called once the enclosing function return, not when it goes out of scope in the for-loop. Is that intended?

Reply:
+1, maybe wrap it in a func() {} to make use of the defer?

Reply:
> Is that intended?

Yes, it is. Each of the `defer` statements is followed by a `return`. A previous design had a single `defer` function which was responsible for flushing the remaining events if the function returns. But there are different requirements for both cases: When returning because the context is marked as done, I want to return the context error. When returning because the channel is closed, I want to return no error. Thus the two different occasions.

Reply:
> +1, maybe wrap it in a func() {} to make use of the defer?

Correct me if I'm wrong but the `defer` is indeed used and does not stack up?!

Reply:
Tbh., now I wonder if we should actually deal with the extra context? Maybe we should obey `ctx` 🤔

Reply:
I refactored the whole method and this shouldn't be a problem anymore. Please re-review.